compiler/rustc_hir_typeck/src/expr.rs

```diff
@@ -2877,7 +2877,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                     // two-phase not needed because index_ty is never mutable
                     self.demand_coerce(idx, idx_t, index_ty, None, AllowTwoPhase::No);
                     self.select_obligations_where_possible(|errors| {
-                        self.point_at_index_if_possible(errors, idx.span)
+                        self.point_at_index(errors, idx.span)
```

Comment:
This makes me pretty uncomfortable. We may be selecting obligations that are only connected to the index expression by long chains of inference, i.e. that are only able to be selected due to coercion that happens above. This seems both non-local and order-dependent.

I would prefer if you just made the heuristic better in `point_at_index_if_possible` -- which is already a hack, but at least currently it's localized to `SliceIndex` trait predicates -- or find out where that `Borrow` obligation comes from below and give it a better `ObligationCauseCode` so that it may be noted more precisely.

Reply:
One check I could make is to pass in the whole expression span in and only modify those that are matched. I initially did that, but then it didn't make any difference to any test in the suite, so I reduced to what there is in the pr.